первый коммит из 3 спринта.  Начало роутинга, маршрут ингридиентов

AlexeySamkov opened the pull request.

AlexeySamkov committed:
первый коммит из 3 спринта.  Начало роутинга, маршрут ингридиентов

AlexeySamkov committed:
промежуточный коммит. написание страниц для работы с авторизацией, и часть логики. в процессе отладки , пока не все работает.

AlexeySamkov committed:
удалил лишнее по совету Андрея. Если захотите воспользоваться куками (лучше не стоит). Точно ну их, что то не получилось с ними

AlexeySamkov committed:
доделал часть по третьему спринту

AlexeySamkov committed:
защита маршрутов

gennady-bars commented on src/services/actions/userActions.jsx:
```diff
@@ -0,0 +1,49 @@
+      const response = await fetchWithRefresh('/auth/user', options);
+      console.log('Успешно получен пользователь', response.user.name);
+      console.log('c Email', response.user.email);
+      dispatch({ type: 'GET_USER_SUCCESS', payload: response });
```
`GET_USER_SUCCESS` и др строки


Это строка

Каждый **тип экшена** должен быть вынесен в **константу** в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const GET_USER_SUCCESS = 'GET_USER_SUCCESS'`

AlexeySamkov replied:
поправил

gennady-bars commented on src/pages/LoginPage/LoginPage.jsx:
```diff
@@ -0,0 +1,75 @@
+import styles from './LoginPage.module.css';
+
+const LoginPage = () => {
+    const [email, setEmail] = useState('');
+    const [password, setPassword] = useState('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

AlexeySamkov replied:
спасибо! код стал чище.

gennady-bars commented on src/pages/LoginPage/LoginPage.jsx:
```diff
@@ -0,0 +1,75 @@
+        <section className={styles.section}>
+            <h2>Вход</h2>         
+            <div className={styles.inputWrapper}>
+                <Input
```
Если есть **инпуты**, то они должны быть внутри тега `form`, и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита.

gennady-bars commented on src/pages/LoginPage/LoginPage.jsx:
```diff
@@ -0,0 +1,75 @@
+                />
+            </div>
+            <div className={styles.inputWrapper}>
+                <Button htmlType="button" type="primary" size="small" extraClass="ml-2" onClick={handleLogin}>
```
**Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**

gennady-bars commented on src/components/ProtectedRouteElement/ProtectedRouteElement.jsx:
```diff
@@ -0,0 +1,20 @@
+import { Navigate, useLocation } from 'react-router-dom';
+import { useSelector } from 'react-redux';
+
+const ProtectedRouteElement = ({ element }) => {
```
**Можно лучше**

Если интересно, то можно сделать вот так универсальный компонент `ProtectedRoute` с переадресацией

```jsx


export default function ProtectedRoute({ children, anonymous = false }) {
  const isLoggedIn = useSelector((store) => store.user.isLoggedIn);

  const location = useLocation();
  const from = location.state?.from || '/';
  // Если разрешен неавторизованный доступ, а пользователь авторизован...
  if (anonymous && isLoggedIn) {
    // ...то отправляем его на предыдущую страницу
    return <Navigate to={ from } />;
  }

  // Если требуется авторизация, а пользователь не авторизован...
  if (!anonymous && !isLoggedIn) {
    // ...то отправляем его на страницу логин
    return <Navigate to="/login" state={{ from: location}}/>;
  }

  // Если все ок, то рендерим внутреннее содержимое
  return children;
}

```

gennady-bars commented on src/components/App/App.jsx:
```diff
@@ -1,38 +1,94 @@
+  const { loading, error, ingredients } = useSelector((state) => state.ingredients);
+  const location = useLocation();
+  const navigate = useNavigate();
+  const background = location.state && location.state.background;
```
Отлично сделана логика отображения попапа и страницы  под ним

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично сделана логика отображения попапа и страницы  под ним


но есть некоторые недочеты:


* Приложение ломается, если открыть попап ингредиента и перезагрузить страницу https://disk.yandex.ru/i/kvna2mbKzNuKDA 
* Нет **автологина**. Если перезагрузить страницу в профиле, то я опять оказываюсь на странице логина, а должен быть автологин и я должен попадать туда, откуда перезагружал страницу
* Каждый **тип экшена** должен быть вынесен в **константу** в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const GET_USER_SUCCESS = 'GET_USER_SUCCESS'`
* Если есть **инпуты**, то они должны быть внутри тега `form`, и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита.
* **Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**

AlexeySamkov committed:
исправил замечания.
привел в порядок экшены, исправил ошибку с текущим ингредиентом, добавил автологин.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
